### DISCUSSION TITLE
Index name error in IndexCreation for generic type class

### DISCUSSION THREAD

**Discussion by ashokgudur** [ravendb/ravendb#13535] [General]
Hi, 

I have the following index definition:

public class GstrDoc_ByReturnPeriod<TDoc> : AbstractIndexCreationTask<TDoc>
                                                where TDoc : GstrHeaderBase
{
        public GstrDoc_ByReturnPeriod()
        {
            Map = docs => docs.Select(doc => new
            {
                FirstName = doc.GstinAccountId,
                LastName = doc.ReturnPeriod
            });
        }
}

_store.ExecuteIndex(new GstrDoc_ByReturnPeriod<TDoc>());

I am getting error: 
"Index name 'GstrDoc/ByReturnPeriod`1' is not permitted. Only letters, digits and characters ('_', '-', '.', '/') are allowed"

Error Stack:

"Raven.Client.Exceptions.RavenException: System.ArgumentException: Index name 'GstrDoc/ByReturnPeriod`1' is not permitted. Only letters, digits and characters ('_', '-', '.', '/') are allowed.\n   at Raven.Server.Documents.Indexes.IndexStore.ValidateStaticIndex(IndexDefinition definition) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Indexes\\IndexStore.cs:line 1012\n   at Raven.Server.Documents.Indexes.IndexStore.CreateIndexInternal(IndexDefinition definition, String raftRequestId, String source) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Indexes\\IndexStore.cs:line 873\n   at Raven.Server.Documents.Handlers.Admin.AdminIndexHandler.PutInternal(Boolean validatedAsAdmin) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Handlers\\Admin\\AdminIndexHandler.cs:line 89\n   at Raven.Server.Documents.Handlers.Admin.AdminIndexHandler.Put() in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Handlers\\Admin\\AdminIndexHandler.cs:line 31\n   at Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Routing\\RequestRouter.cs:line 356\n   at Raven.Server.RavenServerStartup.RequestHandler(HttpContext context) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\RavenServerStartup.cs:line 239 ---> System.ArgumentException: System.ArgumentException: Index name 'GstrDoc/ByReturnPeriod`1' is not permitted. Only letters, digits and characters ('_', '-', '.', '/') are allowed.\n   at Raven.Server.Documents.Indexes.IndexStore.ValidateStaticIndex(IndexDefinition definition) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Indexes\\IndexStore.cs:line 1012\n   at Raven.Server.Documents.Indexes.IndexStore.CreateIndexInternal(IndexDefinition definition, String raftRequestId, String source) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Indexes\\IndexStore.cs:line 873\n   at Raven.Server.Documents.Handlers.Admin.AdminIndexHandler.PutInternal(Boolean validatedAsAdmin) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Handlers\\Admin\\AdminIndexHandler.cs:line 89\n   at Raven.Server.Documents.Handlers.Admin.AdminIndexHandler.Put() in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Documents\\Handlers\\Admin\\AdminIndexHandler.cs:line 31\n   at Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\Routing\\RequestRouter.cs:line 356\n   at Raven.Server.RavenServerStartup.RequestHandler(HttpContext context) in C:\\Builds\\RavenDB-Stable-5.2\\52021\\src\\Raven.Server\\RavenServerStartup.cs:line 239\r\n   --- End of inner exception stack trace ---\r\n   at Raven.Client.Exceptions.ExceptionDispatcher.<Throw>d__2.MoveNext() in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Exceptions\\ExceptionDispatcher.cs:line 110\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Raven.Client.Http.RequestExecutor.<HandleUnsuccessfulResponse>d__138`1.MoveNext() in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Http\\RequestExecutor.cs:line 1454\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Raven.Client.Http.RequestExecutor.<ExecuteAsync>d__114`1.MoveNext() in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Http\\RequestExecutor.cs:line 872\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at Raven.Client.Http.RequestExecutor.<ExecuteAsync>d__114`1.MoveNext() in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Http\\RequestExecutor.cs:line 898\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Raven.Client.Documents.Operations.MaintenanceOperationExecutor.<SendAsync>d__13`1.MoveNext() in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Documents\\Operations\\MaintenanceOperationExecutor.cs:line 62\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Raven.Client.Util.AsyncHelpers.<>c__DisplayClass3_0.<<RunSync>b__0>d.MoveNext() in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Util\\AsyncHelpers.cs:line 59\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   at Raven.Client.Util.AsyncHelpers.RunSync(Func`1 task) in C:\\Builds\\RavenDB-Stable-5.3\\53001\\src\\Raven.Client\\Util\\AsyncHelpers.cs:line 72\r\n   at Aqume.Carbon.Dao.GstrDocDao`1.Delete(String gstinAccountId, String returnPeriod) in C:\\Users\\ashok\\source\\repos\\aqume.com\\Backend\\WebApi\\Carbon\\aqume.carbon.api\\src\\aqume.carbon.api\\Aqume.Carbon.Dao\\GstrDocDao.cs:line 153\r\n   at Aqume.Carbon.Logic.SaveGstr2b2bDataService.Save() in C:\\Users\\ashok\\source\\repos\\aqume.com\\Backend\\WebApi\\Carbon\\aqume.carbon.api\\src\\aqume.carbon.api\\Aqume.Carbon.Logic\\SaveGstr2b2bDataService.cs:line 50\r\n   at Aqume.Carbon.Logic.SaveGstr2b2bDataService.Execute() in C:\\Users\\ashok\\source\\repos\\aqume.com\\Backend\\WebApi\\Carbon\\aqume.carbon.api\\src\\aqume.carbon.api\\Aqume.Carbon.Logic\\SaveGstr2b2bDataService.cs:line 32\r\n   at Aqume.Api.Shared.Helpers.ApiActionExecutor.Execute(Assembly assembly, ApiAction`1 request) in C:\\Users\\ashok\\source\\repos\\aqume.com\\Backend\\WebApi\\Shared\\aqume.api.shared\\src\\aqume.api.shared\\Aqume.Api.Shared\\Helpers\\ApiActionExecutor.cs:line 13\r\n   at Aqume.Carbon.Api.Controllers.ServiceApiController.ExecuteAction(ApiAction`1 req) in C:\\Users\\ashok\\source\\repos\\aqume.com\\Backend\\WebApi\\Carbon\\aqume.carbon.api\\src\\aqume.carbon.api\\Aqume.Carbon.Api\\Controllers\\ServiceApiController.cs:line 44"

**Comment by arekpalinski** [comment]
For sure it's because of \` character in index name "GstrDoc/ByReturnPeriod`1". 

You have some generics around, right? I see you have `where TDoc : GstrHeaderBase`. You you could show us full classes then we could help.

**Comment by ashokgudur** [comment]
Yes, my domain entity class is a generic class as given below the model.

public class Gstr2b2bForm : GstrFormBase<GstGstr2b2bData>
{
    public bool ShouldReplace { get; set; }
    public DateTimeOffset ProcessedAt { get; set; }
}

public class Gstr2b2baForm : GstrFormBase<GstGstr2b2bData>
{
    public bool ShouldReplace { get; set; }
    public DateTimeOffset ProcessedAt { get; set; }
}

public class Gstr2cdnForm : GstrFormBase<GstGstr2b2bData>
{
    public bool ShouldReplace { get; set; }
    public DateTimeOffset ProcessedAt { get; set; }
}

public abstract class GstrFormBase<TSectionData> : DocumentBase
{
    public string UserId { get; set; }
    public string CompanyId { get; set; }
    public string GstinAccountId { get; set; }
    public string ReturnPeriod { get; set; }
    public TSectionData Data { get; set; }
}    

public abstract class DocumentBase
{
    public string Id
}

I'm trying to create below index.
public class GstrDoc_ByReturnPeriod<TDoc> : AbstractIndexCreationTask<TDoc>
                                    where TDoc : GstrHeaderBase
{
    public GstrDoc_ByReturnPeriod()
    {
        Map = docs => docs.Select(doc => new
        {
            GstinAccountId = doc.GstinAccountId,
            ReturnPeriod = doc.ReturnPeriod
        });
    }
}

Since the index name is taken internally from the class name, and it is the generic class, hence the name automatically becomes 'GstrDoc/ByReturnPeriod1' 

Can I give the index name myself? How?

**Reply by ayende** [reply]
Yes, you can override the `IndexName` property
